pkg/operation/botanist/component/extensions/operatingsystemconfig/executor/templates/scripts/execute-cloud-config.tpl.sh

```diff
@@ -171,18 +171,24 @@ else
   rm -f "{{ .pathBootstrapToken }}"
 fi
 
-NODENAME=
+# compare CCD script checksum with node annotation if present
+{{`NODENAME="$(`}}{{ .pathBinaries }}{{`/kubectl --kubeconfig="`}}{{ .pathKubeletKubeconfigReal }}{{`" get node -l "kubernetes.io/hostname=$(hostname)" -o go-template="{{ if .items }}{{ (index .items 0).metadata.name }}{{ end }}")"`}}
+if [[ ! -z "$NODENAME" ]]; then
+  ANNOTATION_CCD_SCRIPT_CHECKSUM=$({{ .pathBinaries }}/kubectl --kubeconfig="{{ .pathKubeletKubeconfigReal }}" get node ${NODENAME} -o jsonpath={'.metadata.annotations.gardener\.cloud\/cloud-config-secret-checksum'})
+  if [[ $(cat ${PATH_CHECKSUM}) != ${ANNOTATION_CCD_SCRIPT_CHECKSUM} ]]; then
+    echo "Actual checksum '$(cat ${PATH_CHECKSUM})' and expected cloud-config secret checksum '${ANNOTATION_CCD_SCRIPT_CHECKSUM}' don't match. Skipping cloud-config evaluation."
+    exit 0
+  else
+    echo "cloud-config secret checksums match. continue evaluating ccd."
+  fi
+fi
```

Comment:
We can absolutely move this down, yes. The difference is that moving it down would allow for restarting systemd services with a node annotation – is this something we do or we do not want? I was under the impression that we wanted to avoid any unnecessary changes to nodes which are about to be removed. If you want me to, I'll move them down.